src/tilesheets.rs

```diff
@@ -292,20 +324,91 @@ impl TilesheetManager {
         let optipng = self
             .tilesheets
             .iter()
-            .map(|tilesheet| {
-                let name = format!("Tilesheet {} {}.png", self.name, tilesheet.size);
-                let path = Path::new(r"work/tilesheets").join(name);
-                save(&tilesheet.img, &path);
-                Command::new("optipng").arg(path).spawn().unwrap()
+            .flat_map(|tilesheet| {
+                tilesheet.layers.iter().enumerate().map(move |(z, layer)| {
+                    let name = format!("Tilesheet {} {} {}.png", self.name, tilesheet.size, z);
+                    let path = BASE_PATH.join(name);
+                    layer.save(&path).unwrap();
+                    Command::new("optipng").arg(path).spawn().unwrap()
+                })
             })
             .collect::<Vec<_>>();
         for mut child in optipng {
             child.wait().unwrap();
         }
     }
     fn upload_sheets(&self) {
-        println!("Tilesheet uploading does not work currently.");
-        println!("Please manually upload the tilesheet images.");
+        println!("Uploading tilesheets.");
+        let token = &self.mw.get_token().unwrap();
+        let failed_uploads = self
+            .tilesheets
+            .iter()
+            .flat_map(|tilesheet| {
+                tilesheet.layers.iter().enumerate().flat_map(move |(z, _)| {
+                    let name = format!("Tilesheet {} {} {}.png", self.name, tilesheet.size, z);
```

Comment:
Computing this path value twice does not make sense. Adding a vector to the `TilesheetManager` struct to hold the paths makes more sense. See [https://github.com/SizableShrimp/ftb-rs/blob/obsolete/src/tilesheets.rs#L307-L324](this example method). Although maybe not implemented the best way possible, it demonstrates how this could be done.